lib/transaction_ops.c

```diff
@@ -53,100 +53,120 @@
  * data type is specified on its edge, i.e string, array, integer, dictionary.
  */
 static int
-trans_find_pkg(struct xbps_handle *xhp, const char *pkg, bool force)
+rpool_find_pkg(struct xbps_handle *xhp, const char *pkg, bool force,
+			   xbps_trans_type_t *ttype, xbps_dictionary_t *pkg_pkgdb, xbps_dictionary_t *pkg_repod)
 {
-	xbps_dictionary_t pkg_pkgdb = NULL, pkg_repod = NULL;
-	xbps_object_t obj;
-	xbps_array_t pkgs;
-	pkg_state_t state = 0;
-	xbps_trans_type_t ttype;
-	const char *repoloc, *repopkgver, *instpkgver, *pkgname;
+	const char *repoloc, *repopkgver, *instpkgver;
 	char buf[XBPS_NAME_SIZE] = {0};
-	int rv = 0;
 
-	assert(pkg != NULL);
+	assert(pkg);
+	assert(ttype);
+	assert(pkg_pkgdb);
+	assert(pkg_repod);
 
 	/*
 	 * Find out if pkg is installed first.
 	 */
 	if (xbps_pkg_name(buf, sizeof(buf), pkg)) {
-		pkg_pkgdb = xbps_pkgdb_get_pkg(xhp, buf);
+		*pkg_pkgdb = xbps_pkgdb_get_pkg(xhp, buf);
```

Comment:
Maybe avoid so many pointer changes by using an intermediate variable and copying from it near the `return 0` line? Not sure it's worth it, but would minimize the diff at the expense of always remembering to copy things out at exit time.